libraries/AP_Scripting/generator/src/main.c

```diff
@@ -2495,16 +2491,19 @@ void emit_type_index_with_operators(struct userdata * data, char * meta_name) {
 
 void emit_loaders(void) {
 
-  emit_type_index_with_operators(parsed_userdata, "userdata");
-  emit_type_index(parsed_singletons, "singleton");
-  emit_type_index(parsed_ap_objects, "ap_object");
+  emit_userdata_type_index(parsed_userdata, "userdata");
+  emit_singleton_type_index(parsed_singletons, "singleton");
+  emit_ap_object_type_index(parsed_ap_objects, "ap_object");
 
   fprintf(source, "static int binding_index(lua_State *L) {\n");
   fprintf(source, "    const char * name = luaL_checkstring(L, 2);\n");
   fprintf(source, "\n");
   fprintf(source, "    bool found = false;\n");
   fprintf(source, "    for (uint32_t i = 0; i < ARRAY_SIZE(singleton_fun); i++) {\n");
   fprintf(source, "        if (strcmp(name, singleton_fun[i].name) == 0) {\n");
+  fprintf(source, "            if (singleton_fun[i].getter && (*singleton_fun[i].getter)() == nullptr) {\n");
```

Comment:
Hum, the assumption is that if its available when it is first called then it will be available at any point afterwards. I'm not sure that is valid in all case, and even if it is now it may not be in the future. 

How about we keep the runtime nulprt check but replace the error string with a generic one. That should keep the a lot of the saving and protect us against the edge case.

Reply:
Most of the Ardupilot code base doesn’t check for null pointers. I almost removed the check entirely except for the possibly unique case of Motors dynamic. The function call overhead is also just large compared to the simple pointer read the singleton load usually gets compiled to.

I will fiddle with this more but possibly bring it back as a separate PR to discuss in more detail.

Reply:
> Most of the Ardupilot code base doesn’t check for null singletons.

If you grep for `get_singleton` I think you will find we do check in the vast majority of cases.

Reply:
I take back my statement about "most", there are definitely a lot of checks. But the pattern is still confusing to me.

Take e.g. `AP_Logger`. Most (but not all) of the 32 calls to `AP_Logger::get_singleton()`, do check for nullptr. However they also all seem to be guarded by `#if HAL_LOGGING_ENABLED` But the singleton is a member variable of AP_Vehicle and AP_Vehicle's subclass is a static variable so it's unclear how it could ever be nullptr.

Additionally, there are 311 calls to `AP::logger()` which returns the singleton without checking for nullptr. Some of these calls are protected by a previous check for nullptr, but certainly not all. Some singleton getter methods on `AP` return pointers, does that mean those are more likely to be nullptr?

Are there any "rules of Ardupilot singletons" anywhere? I don't really know what the developer intended guarantees are for AP_Logger, much less all of them. If I want to use the logger in a new feature, which of the three ways should I do it?

Reply:
I'm probably not the person to answer that. I think we need to keep to the existing pattern here in scripting for now. As you say, if we were going to make it consistent across all singleton users then we could also remove the checks here. But that is outside the scope of this PR.

Reply:
I took your suggestion of simplifying the error message, we can come back to singletons later.

Reply:
Sorry, I what I had in mind was that it was OK to have a simple error message if you were doing the verbose check on first load.

Although I guess this might not be too painful. Let me see what the error look like. 

```
Lua: ./scripts/MotorMatrix_setup.lua:4: calling 'add_motor_raw' on bad self (singleton not available)
```
I think its OK. That should be enough to give a bit of a clue. Although not as good as what we had. I guess wait and see how much flash we get ;)

Reply:
Removed the changes here entirely, couldn't come up with anything that would preserve the message and save a meaningful amount of flash.